Fix ctrl+c in pause module and make other improvements

samdoran opened the pull request:
##### SUMMARY

Fixes #35372

Ctrl + C works in all scenarios.

Use `curses` to control stdout to avoid displaying control characters on screen and make interactive input more user friendly.

Validate the `echo` parameter and ensure it is always cast as a `bool`.

Add integration tests using `pexpect`.

Supercedes #38049 

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request

##### COMPONENT NAME
`pause.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.6
```

bcoca committed:
fix all cases with pause and ctrl+c

now ctrl+c works in most scenarios:
 - naked:
	- pause:
 - with prompt
	- pause: prompt=hi
 - time wait
	- pause: seconds=60
 - time wait with prompt
	- pause: seconds=60 prompt=hi

removed redundant seconds if

fixes #35372

samdoran committed:
Go back to original approach with minor improvements

abadger committed:
Use curses to control stdout

samdoran committed:
Use curses to clear lines on interactive input

samdoran committed:
Validate input for echo parameter and fail nicely if invalid

samdoran committed:
Add integration tests for pause module using pexpect

mattclay commented on test/integration/targets/pause/runme.sh:
```diff
@@ -0,0 +1,25 @@
+
+export ANSIBLE_ROLES_PATH=./roles
+
+python runme.py -i ../../inventory "$@"
```
You shouldn't need a `runme.sh` just to run `runme.py`, since ansible-test will execute `runme.py` directly.

mattclay replied:
However, to avoid having to install `pexpect` for every integration test run via `integration.txt`, it may be better to keep this for now and install `pexepect` via pip here rather than using `integration.txt`.

samdoran replied:
Thanks for answering that! I was going to ask you about that today but you beat me to it.

samdoran committed:
Remove pexpect from integration.txt

samdoran committed:
Install pexpect via runme.sh

PEP8 fixes in test-pause.py

samdoran commented:
@mattclay One thing I've noticed is that inside the container I don't have a tty to query for the appropriate `VERASE`. Would there be a way to add a tty to the container by running it with `-it` or `-t`?


```
Traceback (most recent call last):
  File "test-pause.py", line 17, in <module>
    backspace = termios.tcgetattr(sys.stdin.fileno())[6][termios.VERASE]
termios.error: (25, 'Inappropriate ioctl for device')
```

samdoran committed:
Fix test errors

Use six from ansible.module_utils.six
Use try except when trying to determien erase sequence to account for lack of TTY in containers

samdoran committed:
Expect failure in pause playbooks

mattclay commented:
@samdoran I'll try adding `-it` when running all test containers to see if it causes any issues.

samdoran committed:
Try expected failure for task name that should fail

samdoran commented:
@mattclay I was able to work around that with a `try` `except` statement that hardcodes a backspace escape sequence and that seems to be working.

The last test problem I need to solve is that I am testing aborting the play interactively, which causes the tests to report failure (well, unstable actually). I thought I put the correct string in the task name, but that doesn't seem to fix it. This may be different than other tests in that I am actually aborting the play, but I'm not sure.

mattclay commented:
@samdoran Your latest changes to use `EXPECTED FAILURE` should fix the unstable test results.

mattclay commented:
It's good that you have a work around, because both `-it` and `-t` cause problems for CI.

samdoran committed:
Also test a standard playbook

samdoran committed:
Add changelog entry

samdoran commented:
Thanks for looking in to it.

samdoran committed:
Improve output validation for regular paus test

Accept two digit precision for pause length in test
Check for seconds when seconds is specificed, minutes when minutes is specified

samdoran committed:
Add test for no TTY mode
